spec.bs

```diff
@@ -1767,23 +1760,62 @@ CORP violation report=] algorithm, as leaving it unfenced may cause a privacy le
 
     * |navigable| is a [=fenced navigable container/fenced navigable=];
 
-    Note: This ensures that *all* navigations inside of a <{fencedframe}> are made with the "<a
-    for="history handling behavior">`replace`</a>" mode, regardless of the initiator.
+      Note: This ensures that *all* navigations inside of a <{fencedframe}> are made with the "<a
+      for="history handling behavior">`replace`</a>" mode, regardless of the initiator.
+
+  <wpt>
+    /fenced-frame/history-back-and-forward-should-not-work-in-fenced-tree.https.html
+    /fenced-frame/history-length-fenced-navigations-replace-do-not-contribute-to-joint.https.html
+    /fenced-frame/history-length-outer-page-navigation-not-reflected-in-fenced.https.html
+  </wpt>
 
   Modify step 8 of the same algorithm to include the following condition:
 
     * <var ignore>sourceDocument</var>'s [=node navigable=] is not a [=fenced navigable container=]
       while at the same time |navigable| is a [=fenced navigable container/fenced navigable=].
 
-    Note: This ensures that embedder-initiated navigations can *never* trigger a fragment navigation
-    inside of a <{fencedframe}>.
+      Note: This ensures that embedder-initiated navigations can *never* trigger a fragment
+      navigation inside of a <{fencedframe}>.
 
   <wpt>
     /fenced-frame/fragment-navigation.https.html
+  </wpt>
+
+  Insert these steps immediately after step 16, the step that goes [=in parallel=]:
+
+    1. If |url| is a [=urn uuid=] and |navigable| is a [=fenced navigable container/fenced
+       navigable=]:
+
+      1. Let |config| be the result of [=fenced frame config mapping/finding a
+         config=] in <var ignore>sourceDocument</var>'s [=node navigable=]'s [=navigable/traversable
+         navigable=]'s [=traversable navigable/fenced frame config mapping=].
+
+         Note: This might "wait" for an arbitrary period of time for the |config| associated with
+         the [=urn uuid=] |url| to be "finalized" in the [=traversable navigable/fenced frame config
+         mapping=]. This is why this step runs [=in parallel=]. This navigation will be canceled by
+         any subsequent embedder-initiated navigations, <span class=allow-2119>should</span> they
+         occur, by the usual mechanism that tracks the [=navigable/ongoing navigation=].
+
+      1. [=Assert=] |config|'s [=fenced frame config/mapped url=]'s [=mapped url/value=] is a
+         [=URL=] whose [=url/scheme=] is "`https`".
+      1. Set |url| to |config|'s [=fenced frame config/mapped url=]'s [=mapped url/value=].
+
+      1. Let |config instance| be the result of [=instantiate a config|instantiating=] |config|.
+
+      1. Set |navigable|'s [=navigable/fenced frame config instance=] to |config instance|.
```

Comment:
Does this handle pending/failed navigations properly? The algorithm this is patching seems equivalent to navigation start, and navigation commit is separate. Not sure if each pending embedder-initiated navigation gets its own navigable. (From the changes above, it seems like a fenced navigable is at least something that sticks around and is navigated multiple times.)

Reply:
> Does this handle pending/failed navigations properly? The algorithm this is patching seems equivalent to navigation start, and navigation commit is separate.

Yeah this is the start of the navigation. After this we'll go into the fetching portion which is the other period of time where we wait for an arbitrarily long period of time. "Commit" is _something like_ after-fetch & after all of the navigation failure checks pass, we finally enter ["loading a document"](https://html.spec.whatwg.org/C#populating-a-session-history-entry:loading-a-document).

And yes a fenced navigable (and normal navigable for that matter) sticks around for as long as the element containing it is connected to the DOM. So my plan was to keep overwriting the navigable's config instance for each embedder-initiated navigation. I guess if the navigation were to fail, the config instance would just be hanging around; I suppose we can consider clearing it but I'm not sure it makes a difference either way.

Reply:
But while the navigation is pending (or if it fails), what happens to the config instance of the existing context? In the implementation it's in the NavigationRequest, which I assume here is like `navigationParams`.

Reply:
Eep. Yep I think you're absolutely right. Let me remove the step that assigns the instantiated config for now — I'll add it back in another PR that adjusts things later in the navigation flow.

Reply:
This is what I meant in one of the spec task doc comments, that maybe we should give each embedder-initiated navigation its own fenced navigable. That seems very easy to spec correctly (be sure that everything is isolated across the navigations, even beyond the fenced frame config instance). But there might be other reasons why we don't want to do that.

Reply:
Saw that, sorry I haven't responded yet. The idea is this: browsing context group would be the right primitive to lay things on top of that have to be truly isolated across navigations that swap BCGs. (This matters more with subsequent commits I'll make that force browsing context group _swaps_ within the fenced navigable for embedder-initiated navigations).

See for example https://github.com/whatwg/html/issues/8563 about `name` accidentally getting stored on navigable and persisting across BCG swaps when it probably shouldn't. I'll respond over on Google docs but maybe we should move the config instance to BCG (or browsing context while we have to support URN iframes).